### PR TITLE
cpu support of LpPool(18)

### PR DESCRIPTION
**Pull request by liqunfu**
Signed-off-by: Liqun Fu <liqfu@microsoft.com>

### Description
To support LpPool (18)

### Motivation and Context
for Ort 1.14 release

### CODE REVIEW

File: onnxruntime/core/providers/cpu/nn/pool_functors.h [hunk 1]
```diff
@@ -377,4 +377,171 @@ struct MaxPool3DTask {
...
+            for (int64_t w = wstart; w < wend; w += dilation_w) {
+              if (math::is_a_ge_zero_and_a_lt_b(w, width)) {
+                const int64_t input_index = h * width + w;
+                y_d[pool_index] += static_cast<T>(std::pow(std::abs(x_d[input_index]), p));
+              }
+            }
+          }
+        }
+        y_d[pool_index] = static_cast<T>(std::pow(y_d[pool_index], 1.0f / p));
+        //std::cout << y_d[pool_index] << "\t";
```
Comment (xadupre) [inline onnxruntime/core/providers/cpu/nn/pool_functors.h:471]: remove

File: onnxruntime/core/providers/cpu/nn/pool_functors.h [hunk 2]
```diff
@@ -377,4 +377,171 @@ struct MaxPool3DTask {
...
+  void operator()(std::ptrdiff_t c) const {
+    const T* x_d = X_data + c * x_step;
+    T* y_d = Y_data + c * y_step;
+    for (int64_t ph = 0; ph < pooled_height; ++ph) {
+      int64_t hstart = ph * stride_h - pads[0];
+      int64_t hend = hstart + kernel_shape[0] * dilation_h;
+      y_d[ph] = 0;
+      for (int64_t h = hstart; h < hend; h += dilation_h) {
+        if (math::is_a_ge_zero_and_a_lt_b(h, height)) {
+          y_d[ph] += static_cast<T>(std::pow(std::abs(x_d[h]), p));
```
Comment (xadupre) [inline onnxruntime/core/providers/cpu/nn/pool_functors.h:412]: std::abs? p is an integer, it should work on negative values or maybe it comes from onnx specifications.

File: onnxruntime/core/providers/cpu/nn/pool_functors.h [hunk 3]
```diff
@@ -377,4 +377,171 @@ struct MaxPool3DTask {
...
+        int64_t wstart = pw * stride_w - pads[1];
+        int64_t wend = wstart + kernel_shape[1] * dilation_w;
+        const int64_t pool_index = ph * pooled_width + pw;
+        y_d[pool_index] = 0;
+        for (int64_t h = hstart; h < hend; h += dilation_h) {
+          if (math::is_a_ge_zero_and_a_lt_b(h, height)) {
+            for (int64_t w = wstart; w < wend; w += dilation_w) {
+              if (math::is_a_ge_zero_and_a_lt_b(w, width)) {
+                const int64_t input_index = h * width + w;
+                y_d[pool_index] += static_cast<T>(std::pow(std::abs(x_d[input_index]), p));
```
Comment (xadupre) [inline onnxruntime/core/providers/cpu/nn/pool_functors.h:465]: I wonder if is worth create an intermediate index to store `h * width` which does not change in the inner loop.

**Comment by edgchen1** [comment]
should we add unit tests for the new functionality?

File: onnxruntime/core/providers/cpu/nn/pool_functors.h (hunk 2 above)
Comment (liqunfu) [reply onnxruntime/core/providers/cpu/nn/pool_functors.h:412]: p is integer in ONNX. However it should not be. PyTorch supports float p values.
see also:
https://github.com/microsoft/onnxruntime/issues/6302

This Pr does not change the computation in ort. it just adds support for ceiling mode.

File: onnxruntime/test/providers/cpu/nn/pool_op_test.cc [hunk 4]
```diff
@@ -1331,6 +1331,23 @@ TEST(PoolTest, LpPool) {
   test.Run();
 }
 
+TEST(PoolTest, LpPoolCeilMode) {
```
Comment (pranavsharma) [inline onnxruntime/test/providers/cpu/nn/pool_op_test.cc:1334]: Can we add more tests?

File: onnxruntime/core/providers/cpu/nn/pool.cc [hunk 5]
```diff
@@ -249,6 +249,81 @@ Status MaxPoolV8::ComputeImpl(OpKernelContext* context) const {
...
+      const int64_t dilation_d = pool_attrs_.dilations[2];
+      RunLoop<LpPool3DTask<T>>(tp, onnxruntime::narrow<size_t>(total_channels),
+                                {X_data, Y_data, x_step, y_step,
+                                 dilation_h, dilation_w, dilation_d, pooled_height, pooled_width,
+                                 pooled_depth, stride_h(), stride_w(), stride_d(), height,
+                                width, depth, kernel_shape, pads, p_});
+      break;
+    }
+    default:
+      return Status(ONNXRUNTIME, INVALID_ARGUMENT, "Unsupported pooling size : ");
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/nn/pool.cc:321]: Would be good to mention the size along with the error msg.

**Review by pranavsharma — COMMENTED** [review]
Have you addressed Xavier's comments?

File: onnxruntime/core/providers/cpu/nn/pool_functors.h (hunk 3 above)
Comment (liqunfu) [reply onnxruntime/core/providers/cpu/nn/pool_functors.h:465]: I am concerned with readability here if additional variables are introduced here.  On the hand, compiler will have this computation optimized.

**Comment by liqunfu** [comment]
> Have you addressed Xavier's comments?

I just responded. Thanks for reminding me.

File: onnxruntime/core/providers/cpu/nn/pool.cc [hunk 6]
```diff
@@ -318,7 +318,7 @@ Status LpPoolV18<T>::Compute(OpKernelContext* context) const {
       break;
     }
     default:
-      return Status(ONNXRUNTIME, INVALID_ARGUMENT, "Unsupported pooling size : ");
+      return Status(ONNXRUNTIME, INVALID_ARGUMENT, "Unsupported kernel dimension : " + std::to_string(kernel_shape.size()));
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/nn/pool.cc:321]: nit: you can also use ORT_MAKE_STATUS

**Review by pranavsharma — APPROVED** [review]

**Pull request merged**